spec/policy/scope_check/scope_check_spec.lua

```diff
@@ -0,0 +1,534 @@
+local ScopeCheckPolicy = require('apicast.policy.scope_check')
+
+describe('Scope check policy', function()
+  local ngx_exit_spy
+  local ngx_say_spy
+
+  setup(function()
+    ngx_exit_spy = spy.on(ngx, 'exit')
+    ngx_say_spy = spy.on(ngx, 'say')
+  end)
+
+  describe('.access', function()
+    describe('whitelist', function()
+      describe('check succeeds', function()
+        it('realm role', function()
+          local scope_check_policy = ScopeCheckPolicy.new({
+            scopes = {
+              {
+                realm_roles = { { name = "aaa" } },
+                resource = "/bbb"
+              }
+            }
+          })
+
+          ngx.var = {
+            uri = '/bbb'
+          }
+
+          local context = {
+            jwt = {
+              realm_access = {
+                roles = { "aaa" }
+              }
+            }
+          }
+
+          assert(scope_check_policy:access(context))
+        end)
+
+        it('client role', function()
+          local scope_check_policy = ScopeCheckPolicy.new({
+            scopes = {
+              {
+                client_roles = { { name = "aaa", client = "ccc" } },
+                resource = "/bbb"
+              }
+            }
+          })
+
+          ngx.var = {
+            uri = '/bbb'
+          }
+
+          local context = {
+            jwt = {
+              resource_access = {
+                ccc = {
+                  roles = { "aaa" }
+                }
+              }
+            }
+          }
+
+          assert(scope_check_policy:access(context))
+        end)
+
+        it('multi roles in policy', function()
+          local scope_check_policy = ScopeCheckPolicy.new({
+            scopes = {
+              {
+                realm_roles = { { name = "ddd" } },
+                client_roles = { { name = "aaa", client = "ccc" } },
+                resource = "/bbb"
+              }
+            }
+          })
+
+          ngx.var = {
+            uri = '/bbb'
+          }
+
+          local context = {
+            jwt = {
+              realm_access = {
+                roles = {"ddd", "eee"}
+              },
+              resource_access = {
+                ccc = {
+                  roles = { "aaa" }
+                }
+              }
+            }
+          }
+
+          assert(scope_check_policy:access(context))
+        end)
+
+        it('wildcard', function()
+          local scope_check_policy = ScopeCheckPolicy.new({
+            scopes = {
+              {
+                client_roles = { { name = "aaa", client = "ccc" } },
+                resource = "/bbb/{wildcard}"
```

Comment:
Actually how the MappingRule patters are defined, this should pass even without the wildcard.

`/bbb/` will match `/bbb/*`. To terminate at the end the pattern would have to include `$` like `/bbb/$` would not match `/bbb/a`, but only `/bbb/`.

Is that desired?

Reply:
Changed.